draft-wang-ppm-differential-privacy.md

```diff
@@ -255,12 +255,58 @@ with the correct noise type.
 > TODO: Chris P: we will mention Prio3SumVec because that's what we use to describe aggregator DP with amplification
 
 
-# Differential Privacy Mechanism {#dp mechanism}
+# Differential Privacy Mechanism {#dp-mechanism}
 
 This section should define a framework to apply different dp noise type to a
 client measurement and/or an aggregation.
 
-> TODO: Junye to spell out the interface of dp mechanisms, "Noise Mechanisms to Accomplish DP"
+Clients and Aggregators can implement various noise mechanisms to accomplish
+the DP guarantees described in {{budget}}. The parameters of the noise
+mechanisms should be computed based on the target DP guarantee, and can be
+passed to the initializers of the noise mechanisms.
+
+We define three methods for the noise mechanism interface `DpMechanism`:
+
+* `add_noise` that adds noise to a piece of data. This can be used by a Client
+  to add noise to its measurement.
+
+* `sample_noise` that allows `DpMechanism` to sample noise directly. This can
+  be used by an Aggregator to sample Aggregator DP noise, which then gets merged
+  with the Aggregator's aggregate share.
+
+* `debias` that allows `DpMechanism` to debias the data due to the bias
+  generated by the noise. Note not all noise will need this functionality.
+  We currently think some local randomizers will need this functionality.
+  For example, Appendix A and Appendix C of {{MJTBCCDFGHJKLLMPPPPPRSWZ22}}
+  describe two types of randomized response local randomizers that will
+  need the debiasing functionality.
+
+~~~
+class DpMechanism:
+    DataType
+    DebiasedDataType
+
+    def add_noise(self, data: DataType) -> DataType:
```

Comment:
Add noise into what, the measurement or the aggregate share? How do we know when to call this?

I think it would be worth splitting this class into `ClientDpMechanism` and `AggregatorDpMechanism`.

Reply:
Currently I was going for both, adding noise into measurement or aggregate share, because the noise itself doesn't know about measurement or aggregate share. Assume we turn this PR into tackling primitives instead after #9 is merged, a noise primitive should just have the flexibility of adding noise to data directly, or sampling noise that allows caller to perform future operations on it.

Reply:
> Currently I was going for both, adding noise into measurement or aggregate share, because the noise itself doesn't know about measurement or aggregate share.

I see. In my opinion these should be split into two different classes, as this results in the simplest, most misuse-resistant API surface. Consider an implementation that incorrectly uses a `DpMechanism` intended for Client-DP as a mechanism for Aggregator-DP: Could this result in the aggregate result not being DP, for some definition of DP? In particular, don't we have to tune the sensitivity differently in each case?

> Assume we turn this PR into tackling primitives instead after #9 is merged, a noise primitive should just have the flexibility of adding noise to data directly, or sampling noise that allows caller to perform future operations on it.

Remember, by "primitive" I mean low-level algorithms used to build DP mechanisms. I think this PR should focus on the interface for the latter (DP mechanisms).

Reply:
The idea is adding noise to measurement and adding noise to aggregate share are fundamentally the same operation, having two classes is an implementation detail. We can also have generic add_noise function that applies to Measurement and AggShare types.

Reply:
@junyechen1996 and I discussed offline and converged on the idea of a single, unified `DpMechanism` interface. It turns out that this is really the only viable choice, as there are mechanisms that involve both client and aggregator adding noise, but for which the client bits and aggreagtor bits aren't secure on their own. Thus they MUST be unified.